refactor modal_aero_wateruptake.F90

tangshuaiqi opened the pull request:
This PR includes refactor of all subroutines in modal_aero_wateruptake.F90. The following subroutines are included:
modal_aero_wateruptake_dr
modal_aero_wateruptake_sub (name changed to modal_aero_wateruptake_wetaer)
modal_aero_kohler
makoh_quartic

The subroutine makoh_cubic is removed as it is not executed.

Commit:
modal_aero_wateruptake: refactor makoh_quartic: remove unnecessary dimension and add _r8 on some numbers

Commit:
modal_aero_wateruptake: refactor makoh_quartic: change single-digit variables into multi-digit

Commit:
modal_aero_wateruptake: refactor makoh_cubic: remove the dimension im; change single-digit variables into multi-digit; set eps as a parameter

Commit:
modal_aero_wateruptake: refactor modal_aero_kohler: remove unnecessary input im and the do loop

Commit:
modal_aero_wateruptake: refactor modal_aero_kohler: input/output have unnecessary dimension. change them into one-dimension

Commit:
modal_aero_wateruptake: refactor modal_aero_kohler: change the indent of the code

Commit:
modal_aero_wateruptake: refactor modal_aero_kohler: separate out subroutine find_real_solution

Commit:
modal_aero_wateruptake: refactor modal_aero_kohler: declare some numbers into parameter. add min_max_bound

Commit:
wateruptake: refactor modal_aero_wateruptake_sub: change single-digit variables into multi-digit. add units for input/output variables

Commit:
wateruptake: remove unused subroutine makoh_cubic and related code

Commit:
modal_aero_wateruptake: refactor modal_aero_kohler: remove unexecuted code and related variables

Commit:
modal_aero_wateruptake: refactor modal_aero_wateruptake_dr: remove optional input variable clear_rh_in that is never used and other unused code

Commit:
modal_aero_wateruptake: refactor modal_aero_wateruptake_dr: remove the middle space in enddo and endif

Commit:
wateruptake: refactor modal_aero_wateruptake_sub: change subroutine name to modal_aero_wateruptake_radius

Commit:
modal_aero_wateruptake: refactor modal_aero_wateruptake_dr: separate subroutine modal_aero_wateruptake_wetdens

Commit:
modal_aero_wateruptake: refactor modal_aero_wateruptake_dr: move some code into subroutine modal_aero_wateruptake_radius

Commit:
modal_aero_wateruptake: refactor modal_aero_wateruptake_dr: modal_aero_wateruptake_rh_clearair

Commit:
modal_aero_wateruptake: refactor modal_aero_wateruptake_dr: separate subroutine modal_aero_wateruptake_dryaer

Commit:
modal_aero_wateruptake: refactor modal_aero_wateruptake_dr: change single-digit variable into multi-digit

Commit:
modal_aero_wateruptake: refactor modal_aero_wateruptake_dr: clean unused variables

Commit:
modal_aero_wateruptake: refactor modal_aero_wateruptake_dr: reorganize code for easier read

Commit:
modal_aero_wateruptake: clean up entire file

Commit:
modal_aero_wateruptake: change the name of subroutine modal_aero_wateruptake_radius (original modal_aero_wateruptake_sub) into modal_aero_wateruptake_wetaer as it now includes calculation of several wet aerosol properties

jeff-cohere commented on components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:
```diff
@@ -299,206 +251,228 @@ subroutine modal_aero_wateruptake_dr(state, pbuf, list_idx_in, dgnumdry_m, dgnum
-      wtrvol)
-
-   do m = 1, nmodes
+            ! C++ porting note: these are output but defined in the module
```
Thanks for this comment!

jeff-cohere reviewed (approved):
Good reorganization and simplification!

pressel reviewed (approved):
This looks good to me! Thanks, @tangshuaiqi!

singhbalwinder commented on components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:
```diff
@@ -56,7 +53,6 @@ module modal_aero_wateruptake
 real(r8), allocatable :: rhdeliques(:)
 real(r8), allocatable :: specdens_1(:)
-!$OMP THREADPRIVATE(maer, hygro, naer, dryvol, drymass, dryrad, wetrad, wetvol, wtrvol, rhcrystal, rhdeliques, specdens_1)
 
```
This OMP directive is important if we want to use threading. I think we should keep it.

tangshuaiqi replied:
Okay, I'll add it back.

tangshuaiqi replied:
These OMP lines are removed again as these variables are now declared as regular variables in the corresponding subroutines

singhbalwinder commented on components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:
```diff
@@ -279,13 +227,17 @@ subroutine modal_aero_wateruptake_dr(state, pbuf, list_idx_in, dgnumdry_m, dgnum
+   call phys_getopts(history_aerosol_out = history_aerosol, &
+                     history_verbose_out = history_verbose)
+   ! loop over all aerosol modes
+   call rad_cnst_get_info(list_idx, nmodes=nmodes)
```
`nmodes` can be obtained from `modal_aero_data`. We should get rid of all `rad_cnst...` calls

tangshuaiqi replied:
not quite sure how this is obtained. there is no `nmodes `in `modal_aero_data.F90 `and in `aero_model.F90` `nmodes `is also defined from `rad_cnst_get_info(0, nmodes=nmodes)`. Can you give more information on it?

singhbalwinder replied:
Sorry, in modal_aero_data.F90, it is named `ntot_amode`. Please follow the code by @brigaudet in ndrop.F90 in [subroutine `loadaer`](https://github.com/eagles-project/e3sm_mam4_refactor/blob/cc6f2b8f9d333e4ce3842ae7a071536b0c52ebd1/components/eam/src/physics/cam/ndrop.F90#L1556), it grabs some indices for modes/species which might be helpful here.

tangshuaiqi replied:
import all mode variables from `modal_aero_data `and remove all `rad_cnst_*` calls.

singhbalwinder commented on components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:
```diff
@@ -194,82 +185,39 @@ subroutine modal_aero_wateruptake_dr(state, pbuf, list_idx_in, dgnumdry_m, dgnum
+   if (present(list_idx_in))     list_idx = list_idx_in
 
    !initialize to an invalid value
    naer(:,:,:)    = huge(1.0_r8)
```
It is not necessary but we can define `real(r8), parameter:: huge_real = huge(huge_real)` and replace all `huge(1.0_r8)` with `huge_real`

tangshuaiqi replied:
These huge(1.0_r8) variables look quite strange to me. I am wondering if I can define them as regular real arrays. Also, I feel these module variables can be just declared as input or output variables in each subroutine. I tried to change the variable specdens_1 and it passes the short-term test. @singhbalwinder how do you think?

singhbalwinder commented on components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:
```diff
@@ -194,82 +185,39 @@ subroutine modal_aero_wateruptake_dr(state, pbuf, list_idx_in, dgnumdry_m, dgnum
-   if (present(list_idx_in)) then
-      list_idx = list_idx_in
-
-      ! check that all optional args for diagnostic mode are present
```
I am not sure why this code was removed. These are checks to ensure that this subroutine has the right set of variables and if they are allocated. I think these should be kept.

tangshuaiqi replied:
I did not remove this code, just shrink two lines into one line.

singhbalwinder replied:
I think I tagged the wrong line. I am talking about the following code:
```
if (.not. present(dgnumdry_m) .or. .not. present(dgnumwet_m) .or. &
          .not. present(qaerwat_m)) then
         call endrun('modal_aero_wateruptake_dr called '// &
              'with list_idx_in but required args not present '//errmsg(__FILE__,__LINE__))
      end if

      ! arrays for diagnostic calculations must be allocated
      if (.not. allocated(dgnumdry_m) .or. .not. allocated(dgnumwet_m) .or. &
          .not. allocated(qaerwat_m)) then
         call endrun('modal_aero_wateruptake_dr called '// &
              'with list_idx_in but required args not allocated '//errmsg(__FILE__,__LINE__))
      end if
   end if ! if present(list_idx_in)
```

tangshuaiqi replied:
got it. I will add these checks back

singhbalwinder commented on components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:
```diff
@@ -299,206 +251,228 @@ subroutine modal_aero_wateruptake_dr(state, pbuf, list_idx_in, dgnumdry_m, dgnum
       dgncur_awet => dgnumwet_m
       qaerwat     => qaerwat_m
       if(present(wetdens_m)) then
-         if (.not. allocated(wetdens_m)) then
```
Same here. I am not sure why this check needs to be removed.

tangshuaiqi replied:
just because this check was never executed. Actually wetdens_m is never input in any calls. I may just remove this input variable.

singhbalwinder replied:
Thanks. Yes, please feel free to remove it if it is not used anywhere.

singhbalwinder commented on components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:
```diff
@@ -299,206 +251,228 @@ subroutine modal_aero_wateruptake_dr(state, pbuf, list_idx_in, dgnumdry_m, dgnum
 
    !----------------------------------------------------------------------------
    ! retreive aerosol properties
+   call modal_aero_wateruptake_dryaer( ncol, nmodes, list_idx,    & ! in
```
We should avoid sending `state` as an argument to this routine. It looks like the `state` is being used to grab mass mixing ratios (mmrs). We should directly get them from the `state%q` array and send a `state_q`. We may need to get the right mmrs using indices from the `modal_aro_data` module.

singhbalwinder commented on components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:
```diff
@@ -299,206 +251,228 @@ subroutine modal_aero_wateruptake_dr(state, pbuf, list_idx_in, dgnumdry_m, dgnum
-   do m = 1, nmodes
+   !----------------------------------------------------------------------------
+   ! estimate clear air relative humidity using cloud fraction
+   h2ommr       => state%q(:,:,1)
```
Can we extract these at the beginning of this subroutine, since `state` is intent-in?

tangshuaiqi replied:
done

singhbalwinder commented on components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:
```diff
@@ -299,206 +251,228 @@ subroutine modal_aero_wateruptake_dr(state, pbuf, list_idx_in, dgnumdry_m, dgnum
-      dryvolmr(:,:) = 0._r8
+   ! get cloud fraction
+   itim_old    =  pbuf_old_tim_idx()
+   call pbuf_get_field(pbuf, cld_idx, cldn, start=(/1,1,itim_old/), kount=(/pcols,pver,1/) )
```
This should also be extracted at the beginning of the subroutine.

tangshuaiqi replied:
done

singhbalwinder commented on components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:
```diff
@@ -299,206 +251,228 @@ subroutine modal_aero_wateruptake_dr(state, pbuf, list_idx_in, dgnumdry_m, dgnum
+      dryvolmr(:,:) = 0._r8
+
+      ! get mode properties
+      call rad_cnst_get_mode_props(list_idx, imode,        & ! in
```
All this info for an `imode` can be grabbed from `modal_aero_data` module. We should avoid using `rad_cnst...` calls.

singhbalwinder commented on components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:
```diff
@@ -299,206 +251,228 @@ subroutine modal_aero_wateruptake_dr(state, pbuf, list_idx_in, dgnumdry_m, dgnum
+                rhdeliques=rhdeliques(imode)               ) ! out
       alnsg = log(sigmag)
+      ! get mode info
+      call rad_cnst_get_info(list_idx, imode, & ! in
```
Same here, all the info can be obtained from `modal_aero_data`.

singhbalwinder commented on components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:
```diff
@@ -299,206 +251,228 @@ subroutine modal_aero_wateruptake_dr(state, pbuf, list_idx_in, dgnumdry_m, dgnum
+   real(r8), parameter  :: cldn_thresh = 1.0_r8      ! threshold cloud fraction
+
+   do kk = top_lev, pver
+      call qsat_water(temperature(:ncol,kk), pmid(:ncol,kk), es(:ncol), qs(:ncol))
```
I guess `es` is never used, so the only output that is used is `qs`. We still need to send `es` as it is not an optional variable.

tangshuaiqi replied:
Yes. I'll keep `es `as it is a required output variable of `qsat_water`.

singhbalwinder reviewed (changes requested):
I liked the way you refactored it! The code is easy to follow and understand. I left some comments about rad_cnst calls and some code rearrangements. Overall looks great!

Commit:
modal_aero_wateruptake: add OMP threading code back

Commit:
modal_aero_wateruptake: move some variable declaration into the beginning of the subroutine

tangshuaiqi commented:
Thank you @singhbalwinder for your suggestions. I incorporated most of the comments but the only question is how to get rid of the rad_cnst_* calls. Not quite sure how to do that.

Commit:
wateruptake: remove unused input variable wetdens_m

Commit:
wateruptake: import mode variables from modal_aero_data and remove all rad_cnst_* calls

Commit:
wateruptake: move assign_la_lc into mam_support.F90

Commit:
wateruptake: avoid input state and pbuf in subroutine and remove unused input variables

Commit:
wateruptake: move all small values at the beginning of the module

Commit:
wateruptake: move all module variables that are allocated as huge array into regular subroutine variables

Commit:
Merge branch 'refactor-maint-2.0' into tangshuaiqi/wateruptake

tangshuaiqi commented:
I moved all the module variables that are allocated as huge() to regular subroutine variables, and removed the related OMP thread lines. All the tests are passed:
  PASS ERP_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS ERS.ne4pg2_oQU480.F2010.compy_intel
  PASS PEM_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS PET.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P32x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_Ly1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS.ne30pg2_EC30to60E2r2.F2010.compy_intel
@singhbalwinder please take a look and approve if there is nothing else need to be revised.

singhbalwinder commented on components/eam/src/chemistry/utils/mam_support.F90:
```diff
@@ -12,10 +12,39 @@ module mam_support
-
+!===============================================================================
+   subroutine assign_la_lc( imode,      ispec,          & ! in
+                            la,         lc              ) ! out
```
This is a nice subroutine that can be used at multiple places. Thanks for adding this!

singhbalwinder reviewed (approved):
Added a minor question about a code that has been removed. Rest everything looks good!

Commit:
wateruptake: add back the checks to ensure subroutine has the right set of variables and if they are allocated.

tangshuaiqi commented:
addressed the remaining comments and passed all tests. merge it to the main branch.
